File: 30daysOFcode_Julia/day 04.jl
mutable struct Person
    age::Int64

    function Person(age::Int64)  
        this = new()
        if age < 0
            println("Age is not valid")
            this.age = 0
            return this
        else
            this.age = age
            return this 
        end
    end
end

function yearPasses(self::Person)
    self.age += 1
end

function amIOld(self::Person)
    if self.age < 13
        println("You are young.")
    elseif self.age < 18
        println("You are a teenager.")
    else
        println("You are old.")
    end
end

# a = Person(111)
# yearPasses(a)
# println(a.age)
# amIOld(a)

File: 30daysOFcode_Rust/day 04.rs
#![allow(non_snake_case)]

struct Person {
    age: i32
}

impl Person {
    fn new(initialAge: i32) -> Person {
        let initial_Age = std::cmp::max(initialAge, 0);
        if initialAge < 0 {
            println!("Age is not valid, setting age to 0.");
        }

        return Person { age: initial_Age };
    }

    fn amIOld(&self) {
        match self.age {
            0..=12 => println!("You are young."),
            13..=17 => println!("You are a teenager."),
            _ => println!("You are old."),
        }
    }

    fn yearPasses(&mut self) {
        self.age += 1;
    }
}

fn main() {
    let T: i32 = read_line().trim().parse().unwrap();
 
    for _ in 0..T {
        let age = read_line().trim().parse().unwrap();
        let mut p = Person::new(age);
 
        p.amIOld();
 
        for _ in 0..3 {
            p.yearPasses();
        }
 
        p.amIOld();
        println!("");
    }
}

fn read_line() -> String {
   let mut  input = String::new();
   std::io::stdin().read_line(&mut input).expect("Could not read stdin!");
    return input;
}
